source/common/config/version_converter.cc

```diff
@@ -65,87 +57,165 @@ void constMessageFieldVisitor(
+}
+
+// Create a new dynamic message based on some message wire cast to the target
+// descriptor. If the descriptor is null, a copy is performed.
+DynamicMessagePtr createForDescriptorWithCast(const Protobuf::Message& message,
+                                              const Protobuf::Descriptor* desc) {
+  auto dynamic_message = std::make_unique<DynamicMessage>();
+  if (desc != nullptr) {
+    dynamic_message->msg_.reset(dynamic_message->dynamic_msg_factory_.GetPrototype(desc)->New());
+    wireCast(message, *dynamic_message->msg_);
+    return dynamic_message;
+  }
+  // Unnecessary copy, since the existing message is being treated as
+  // "dynamic". However, we want to transfer an owned object, so this is the
+  // best we can do.
+  dynamic_message->msg_.reset(message.New());
+  dynamic_message->msg_->MergeFrom(message);
+  return dynamic_message;
+}
+
+// This needs to be recursive, since sub-messages are consumed and stored
+// internally, we later want to recover their original types.
+void annotateWithOriginalType(const Protobuf::Descriptor& prev_descriptor,
+                              Protobuf::Message& next_message) {
+  class TypeAnnotatingProtoVisitor : public ProtoVisitor {
+  public:
+    void onMessage(Protobuf::Message& message, const void* ctxt) override {
+      const Protobuf::Descriptor* descriptor = message.GetDescriptor();
+      const Protobuf::Reflection* reflection = message.GetReflection();
+      const Protobuf::Descriptor& prev_descriptor = *static_cast<const Protobuf::Descriptor*>(ctxt);
+      // If they are the same type, there's no possibility of any different type
+      // further down, so we're done.
+      if (descriptor->full_name() == prev_descriptor.full_name()) {
+        return;
+      }
+      auto* unknown_field_set = reflection->MutableUnknownFields(&message);
+      unknown_field_set->AddLengthDelimited(ProtobufWellKnown::OriginalTypeFieldNumber,
+                                            prev_descriptor.full_name());
+    }
+
+    const void* onField(Protobuf::Message&, const Protobuf::FieldDescriptor& field,
+                        const void* ctxt) override {
+      const Protobuf::Descriptor& prev_descriptor = *static_cast<const Protobuf::Descriptor*>(ctxt);
+      // TODO(htuch): This is a terrible hack, there should be no per-resource
+      // business logic in this file. The reason this is required is that
+      // endpoints, when captured in configuration such as inlined hosts in
+      // Clusters for config dump purposes, can potentially contribute a
+      // significant amount to memory consumption. stats_integration_test
+      // complains as a result if we increase any memory due to type annotations.
+      // In theory, we should be able to just clean up these annotations in
+      // ClusterManagerImpl with type erasure, but protobuf doesn't free up memory
+      // as expected, we probably need some arena level trick to address this.
```

Comment:
Sorry I'm not 100% following what the problem is and what we are doing here? It looks like this causes the prev <-> current comparison to be true? And why does this matter?

Reply:
The issue is that tagging original type information on endpoints is really expensive memory-wise, since we may have lots of endpoints and endpoints themselves are tiny, i.e. just addresses in some cases. So, the overhead is likely unacceptable for these of supporting original message recovery. In addition, we don't (today) support config dumping of endpoints directly, so it seems to make sense to just bypass adding this overhead.

This is only there because of `stats_integration_test`, which complained about the extra overhead, it noted 1763 per endpoint, vs the previous bound of 1350, so ~30% extra overhead. For clusters, it's much less, since they have other overheads (you can see the updates to that test in this file).

The solution in this PR is to special case endpoints and avoid ever annotating with original type information.